### ISSUE TITLE
Demo version support

### ISSUE THREAD

**Issue by ghost** [vcmi/vcmi#2126]
**Describe your proposal**
I would like to be able to play the demo version of HoMM3 with VCMI.

**Additional context**
The demo can be downloaded from here https://www.gamefront.com/files/heroes-of-might-magic-iii-the-restoration-of-erathia-demo-english/download
The file we're looking for is `h3demo.exe` with size of `98265058` bytes and sha256 sum `83cbb6319386d6deba65273f53a60395fda8cce74cc24177b2ac3560ded9c315`.
To get from the downloaded file to HoMM3 gamedata, 2 extractions need to be made:
1. Extract data1.cab and data1.hdr from the installer: `7z e h3demo.exe /Disk1/data1.cab /Disk1/data1.hdr`
2. Extract gamedata from data1.cab:
```
unshield x data1.cab H3demo.h3m StrongHold.mp3 'Surrender Battle.mp3' 'Win Scenario.mp3' 'Retreat Battle.mp3' UltimateLose.mp3 LoseCombat.mp3 'Defend Castle.mp3' 'Win Battle.mp3' LoseCastle.mp3HEROES3.SND H3BITMAP.LOD H3SPRITE.LOD VIDEO.VID
```
3. Your `./Program_Files` directory is now a suitable demo asset directory. VCMI should be able to pick this up if you rename it to let's say `~/.local/share/vcmi`
Once this is done, the game should start.
However, 1.2.1 does not start and this is what happens:
```
Starting... 
VCMI 1.2.1 (client)
Creating console and configuring logger: 0 ms                                                                                                                                                                                                
The log file will be saved to "/home/kerneliusz/.cache/vcmi/VCMI_Client_log.txt"                                                                                                                                                                   
        Initialization: 2 ms                                                                                                                                                                                                                 
        Data loading: 13 ms                                                                                                                                                                                                                  
        Mod handler: 0 ms                                                                                                                                                                                                                    
        Mod filesystems: 11 ms                                                                                                                                                                                                               
Initialized logging system based on settings successfully.                                                                                                                                                                                   
[log level] ai => not set                                                                                                                                                                                                                    
[log level] animation => not set                                                                                                                                                                                                             
[log level] bonus => not set                                                                                                                                                                                                                 
[log level] global => trace                                                                                                                                                                                                                  
[log level] mod => not set                                                                                                                                                                                                                   
[log level] network => not set                                                                                                                                                                                                               
Error: campaign movies was not found!                                                                                                                                                                                                        
Error: campaign music was not found!                                                                                                                                                                                                         
Loading settings: 30 ms                                                                                                                                                                                                                      
Found 3 render drivers                                                                                                                                                                                                                       
        opengl (active)                                                                                                                                                                                                                      
        opengles2                                                                                                                                                                                                                            
        software                                                                                                                                                                                                                             
Check display mode: requested 800 x 600; available up to 1920 x 1080                                                                                                                                                                         
Set logical rendering resolution to 800x600                                                                                                                                                                                                  
Created renderer opengl                                                                                                                                                                                                                      
        Initializing screen: 80 ms                                                                                                                                                                                                           
        Initializing video: 0 ms                                                                                                                                                                                                             
Initializing screen and sound handling: 2 ms                                                                                                                                                                                                 
                 Game Settings handler: 0 ms                                                                                                                                                                                                 
terminate called after throwing an instance of 'std::runtime_error'                                                                                                                                                                          
  what():  Resource with name DATA/TENTCOLR and type TEXT wasn't found.
Przerwane
```

The reason why I would like this to be a working feature is because you can freely distribute this h3demo.exe file, and that makes it legal to redistribute VCMI with free data to make it possible for everyone to try VCMI even if they don't own the full version of HoMM3.

**Comment by kerneliusz** [comment]
I found another issue and apparently there exists a Mac OS 9 demo at https://archive.org/download/heroes_3_complete_demo/heroes_3_complete_demo.sit

**Comment by krs0** [comment]
fheroes2 does it and installing it is very easy.

**Comment by kambala-decapitator** [comment]
> there exists a Mac OS 9 demo

In case you think that someone can test this demo, MacOS 9 is like 20 years old and you can be run only in emulator as it supports only PowerPC CPU (modern Macs that run Mac OS X / macOS have Intel / Apple Silicon CPU)

**Comment by kerneliusz** [comment]
The demo isn't just executables - it's also gamedata and that is what VCMI should support. From what I've seen in the console logs VCMI makes too many assumptions that it's gonna run the full version of SOD for this to work.

Being able to run the demo as is - without VCMI - should IMHO only be done to see if VCMI is correctly implementing HoMM3.

**Comment by Nordsoft91** [comment]
@kerneliusz , you can check out support of RoE demo from this PR
https://github.com/vcmi/vcmi/actions/runs/4901196735

To play correctly, you need to install this mode.
[any-game-data.zip](https://github.com/vcmi/vcmi/files/11414141/any-game-data.zip)

**Comment by kerneliusz** [comment]
@Nordsoft91 many thanks! I'll scrupulously document my experience so far, so please forgive for the tl;dr.

So I've checked out sha commit `823d4b655c58c1ef4617b8b7ff538accfe4ba7d6` from https://github.com/Nordsoft91/vcmi/tree/demo-support and I downloaded https://github.com/vcmi/vcmi/files/11414141/any-game-data.zip to `~/.local/share/vcmi/Mods`.
For the record it can be done like so:
```
pushd ~/.local/share/vcmi/Mods
curl -LO https://github.com/vcmi/vcmi/files/11414141/any-game-data.zip
unzip any-game-data.zip 'any-game-data/*' -x '*/.DS_Store'
rm any-game-data.zip
popd
```
That way, we will only extract the mod directory without `__MACOSX` dir and without `.DS_Store` files scattered all over the zip file.
Your config file should look like this after `vcmilauncher` has successfully ingested any-game-data mod (for the record this is the `~/.config/vcmi/modSettings.json` file):
```
{
        "activeMods" : {
                "any-game-data" : {
                        "active" : true,
                        "checksum" : "00000000",
                        "mods" : {
                                "ab-data" : {
                                        "active" : true,
                                        "checksum" : "00000000",
                                        "validated" : true
                                },
                                "demo-data" : {
                                        "active" : true,
                                        "checksum" : "00000000",
                                        "mods" : {
                                                "demo-menu" : {
                                                        "active" : true,
                                                        "checksum" : "00000000",
                                                        "validated" : true
                                                }
                                        },
                                        "validated" : true
                                },
                                "roe-data" : {
                                        "active" : true,
                                        "checksum" : "00000000",
                                        "validated" : true
                                }
                        },
                        "validated" : true
                },
                "vcmi" : {
                        "active" : true,
                        "checksum" : "794b548f",
                        "validated" : true
                }
        },
        "core" : {
                "active" : true,
                "checksum" : "b984b89f",
                "validated" : false
        }
}
```
Now unfortunately, running with demo assets from Heroes 3 Complete for Mac OS 9 is pretty janky in a bad sense:
1. We replicate this gem no problem https://github.com/vcmi/vcmi/issues/1198#issue-1476525481
![image](https://user-images.githubusercontent.com/20668811/236869246-c5c02789-04f8-4c77-801a-e5a027660c99.png)
2. The border for the window that pops up when VCMI starts a server is also affected by this bug:
![image](https://user-images.githubusercontent.com/20668811/236869573-67a7ed48-a2f6-4b52-b0d8-78ef279fb8ac.png)
3. Trying to pick up the Thunder Helmet from inventory crashed the game with following log: 
[VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11422724/VCMI_Client_log.txt)
I'll give it another go in a while with Windows ROE demo.

**Comment by kerneliusz** [comment]
Starting the game with the Windows demo of ROE got us a completely different flavor of bugs.
1. First of all the game wouldn't start without installing `vcmi-extras` and `vcmi-extras.arrowtowericons` mods.
2. We can reproduce the server window from https://github.com/vcmi/vcmi/issues/2126#issuecomment-1538652147 not getting its own window - see screenshot 2 from that post.
3. Clicking "BEGIN" will just make the game go back and give us console output that suggests we check the server logs because apparently the server crashed.
4. The menu is still janky, but less:
![image](https://user-images.githubusercontent.com/20668811/236875397-50c71a0d-a908-42d9-907c-073860da4716.png)

The logs:
1. [VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11422820/VCMI_Client_log.txt)
2. [server_log.txt](https://github.com/vcmi/vcmi/files/11422825/server_log.txt)

**Comment by Nordsoft91** [comment]
@kerneliusz , hm, I believe this part from `~/.config/vcmi/modSettings.json` is not quite correct.
You should enable only one of submods, in this case "demo-data" I believe (for ROE-demo).
Rest should be disabled (e.g `"active": false`)

Yes, mods themselves are not production ready, also, I didn't upgrade mod specicly for macOS-demo (needs upgrade to solve https://github.com/vcmi/vcmi/issues/1198#issue-1476525481).

Those issues are mostly about mod incompletion, I was planning to have some image-mapping or add custom graphics.
If you'd like to, you can continue working on those mods.

**Comment by kerneliusz** [comment]
@Nordsoft91 what if I just fixed all the wrong assumptions that VCMI makes?

I'm trying to keep a list of things that are obviously going wrong and so far:
1. It assumes it has access to ROE, AB and SOD gamedata, instead of what the demos limit the game to - 3 castles and a small choice of artifacts.
2. It assumes something about tent colors which is fixed by the mods.
3. It assumes that all music files have an extension to their filename and that extension is `.mp3`. This is demonstrably false because the Mac demo delivers music in `IFF data, AIFF-C compressed audio` format without any filename extension. ffmpeg recognizes this as `Stream #0:0: Audio: adpcm_ima_qt (ima4 / 0x34616D69), 22050 Hz, 2 channels, s16p, 176 kb/s`.
4. There are parts of the ROE+AB+SOD soundtrack missing in the demos and VCMI has no idea how to handle this so instead it just doesn't play the music it can't find.

**Comment by Nordsoft91** [comment]
@kerneliusz , yes, actually not only about castles. RoE has less heroes, arts, creatures, map objects. 2 is correct. 3, 4 - good point, because I didn't test music and sounds. Usually, music isn't started if it missing. AFAIK, we support mp3 and ogg formats but not sure about IFF.

**Comment by Laserlicht** [comment]
In principle, the MAC version (complete) has more resources. Interestingly, there are also in the Win version (RoE) a few files that are missing in the MAC version:
- More music (and in MP3 format)
- _I've don't compared Sounds & Video_
- following files (many creatures & puzzle elements):

```
 +h3bitmap.lod            24.828.818 
 |<<  |+altarart.png              184.139 
 |<<  |+credits.png               836.265 
 |<<  |+gamselb0.png              495.571 
 |<<  |+heroscr3.png              213.966 
 |<<  |+loadgame.png               28.734 
 |<<  |+puzdun00.png                6.514 
 |<<  |+puzdun01.png                6.056 
 |<<  |+puzdun02.png               10.107 
 |<<  |+puzdun03.png                8.920 
 |<<  |+puzdun04.png                4.313 
 |<<  |+puzdun05.png                9.932 
 |<<  |+puzdun06.png                5.139 
 |<<  |+puzdun07.png                6.601 
 |<<  |+puzdun08.png                4.682 
 |<<  |+puzdun09.png                6.828 
 |<<  |+puzdun10.png                8.721 
 |<<  |+puzdun11.png                6.038 
 |<<  |+puzdun12.png                6.743 
 |<<  |+puzdun13.png                9.045 
 |<<  |+puzdun14.png                7.874 
 |<<  |+puzdun15.png                6.423 
 |<<  |+puzdun16.png                5.042 
 |<<  |+puzdun17.png                6.167 
 |<<  |+puzdun18.png                3.817 
 |<<  |+puzdun19.png                6.989 
 |<<  |+puzdun20.png                6.082 
 |<<  |+puzdun21.png                5.426 
 |<<  |+puzdun22.png                5.887 
 |<<  |+puzdun23.png                4.685 
 |<<  |+puzdun24.png                6.673 
 |<<  |+puzdun25.png                8.511 
 |<<  |+puzdun26.png                5.130 
 |<<  |+puzdun27.png                6.371 
 |<<  |+puzdun28.png                5.952 
 |<<  |+puzdun29.png                3.521 
 |<<  |+puzdun30.png                7.302 
 |<<  |+puzdun31.png                3.951 
 |<<  |+puzdun32.png                3.472 
 |<<  |+puzdun33.png                6.282 
 |<<  |+puzdun34.png                8.558 
 |<<  |+puzdun35.png                3.567 
 |<<  |+puzdun36.png                4.547 
 |<<  |+puzdun37.png                7.568 
 |<<  |+puzdun38.png                5.346 
 |<<  |+puzdun39.png                4.413 
 |<<  |+puzdun40.png                4.249 
 |<<  |+puzdun41.png                8.079 
 |<<  |+puzdun42.png                8.521 
 |<<  |+puzdun43.png                6.153 
 |<<  |+puzdun44.png                3.526 
 |<<  |+puzdun45.png                6.849 
 |<<  |+puzdun46.png                6.145 
 |<<  |+puzdun47.png                4.785 
 |<<  |+puznec00.png                7.724 
 |<<  |+puznec01.png                5.143 
 |<<  |+puznec02.png                3.859 
 |<<  |+puznec03.png                4.742 
 |<<  |+puznec04.png                6.995 
 |<<  |+puznec05.png                4.733 
 |<<  |+puznec06.png                3.954 
 |<<  |+puznec07.png                6.580 
 |<<  |+puznec08.png                4.228 
 |<<  |+puznec09.png                7.681 
 |<<  |+puznec10.png                5.267 
 |<<  |+puznec11.png                3.961 
 |<<  |+puznec12.png                6.401 
 |<<  |+puznec13.png                8.563 
 |<<  |+puznec14.png                9.471 
 |<<  |+puznec15.png                6.084 
 |<<  |+puznec16.png               10.318 
 |<<  |+puznec17.png                8.938 
 |<<  |+puznec18.png                3.888 
 |<<  |+puznec19.png                5.077 
 |<<  |+puznec20.png                7.899 
 |<<  |+puznec21.png                9.148 
 |<<  |+puznec22.png                4.642 
 |<<  |+puznec23.png                7.474 
 |<<  |+puznec24.png                4.912 
 |<<  |+puznec25.png                7.131 
 |<<  |+puznec26.png                7.819 
 |<<  |+puznec27.png                5.568 
 |<<  |+puznec28.png                5.867 
 |<<  |+puznec29.png                8.130 
 |<<  |+puznec30.png                2.093 
 |<<  |+puznec31.png                5.510 
 |<<  |+puznec32.png                6.047 
 |<<  |+puznec33.png                3.983 
 |<<  |+puznec34.png                8.303 
 |<<  |+puznec35.png                4.472 
 |<<  |+puznec36.png                6.402 
 |<<  |+puznec37.png                4.355 
 |<<  |+puznec38.png                4.260 
 |<<  |+puznec39.png                6.447 
 |<<  |+puznec40.png                5.575 
 |<<  |+puznec41.png                5.215 
 |<<  |+puznec42.png                8.508 
 |<<  |+puznec43.png                6.592 
 |<<  |+puznec44.png                5.837 
 |<<  |+puznec45.png                2.961 
 |<<  |+puznec46.png                3.808 
 |<<  |+puznec47.png                3.615 
 |<<  |+tpcasfor.png               12.002 
 |<<  |+tpcasinf.png               10.107 
 |<<  |+tpcasram.png               10.568 
 |<<  |+tpcasstr.png               10.575 
 |<<  |+tpthbktw.png              149.602 
 +h3sprite.lod            80.621.087 
 |<<  |+cabehe.def                460.525 
 |<<  |+cadevl.def                425.972 
 |<<  |+camage.def                299.576 
 |<<  |+capegs.def                341.467 
 |<<  |+cbasil.def                274.121 
 |<<  |+cbgog.def                 278.065 
 |<<  |+cbtree.def                275.131 
 |<<  |+ccerbu.def                311.976 
 |<<  |+ccgorg.def                241.479 
 |<<  |+cchydr.def                402.068 
 |<<  |+cddrag.def                800.452 
 |<<  |+cdevil.def                402.716 
 |<<  |+cdgole.def                155.869 
 |<<  |+cdrfir.def                122.977 
 |<<  |+cefree.def                176.283 
 |<<  |+cefres.def                176.910 
 |<<  |+cgbasi.def                319.998 
 |<<  |+cgdrag.def                721.141 
 |<<  |+cgenie.def                193.214 
 |<<  |+cggole.def                193.054 
 |<<  |+cgrelf.def                229.730 
 |<<  |+cgtita.def                312.700 
 |<<  |+chydra.def                328.252 
 |<<  |+cigole.def                162.872 
 |<<  |+cltita.def                224.687 
 |<<  |+cmage.def                 238.938 
 |<<  |+cnaga.def                 247.266 
 |<<  |+cnagag.def                282.024 
 |<<  |+cogmag.def                263.370 
 |<<  |+cogre.def                 197.915 
 |<<  |+cohdem.def                201.631 
 |<<  |+corcch.def                255.396 
 |<<  |+cpfien.def                219.139 
 |<<  |+cpfoe.def                 239.632 
 |<<  |+croc.def                  553.737 
 |<<  |+csulta.def                218.001 
 |<<  |+ctbird.def                569.887 
 |<<  |+cthdem.def                216.351 
 |<<  |+ctree.def                 226.905 
 |<<  |+cunico.def                264.117 
 |<<  |+cwunic.def                352.826 
 |<<  |+cwyver.def                351.186 
 |<<  |+cwyvmn.def                418.425 
 |<<  |+cybehe.def                427.668 
```

So if you want to have as many "free" resources as possible, you have to combine both. For this purpose (and for comparison) I have made a small script for downloading and extracting (as a Jupyter notebook): https://github.com/Laserlicht/h3_demo_extract

**Comment by GeorgeK1ng** [comment]
VCMI can be used with merged demo files created using the [Heroes of Might and Magic III Demo Builder for VCMI](https://github.com/GeorgeK1ng/heroes3demo), along with the [Heroes III Demo](https://github.com/vcmi-mods/demo-support) and [VCMI Extras](https://github.com/vcmi-mods/vcmi-extras) mods enabled.

No bugs or issues – just enjoy limited gameplay with Castle, Dungeon, and Necropolis!  
There are no further plans for expanding demo support directly to VCMI.
